### PR TITLE
fix(Windows): ensure both Windows Powershell and Powershell Core are available

### PR DESCRIPTION
**Pull request by lemeurherve**
Ref: jenkins-infra/helpdesk#2974, more specifically https://github.com/jenkins-infra/helpdesk/issues/2974#issuecomment-1161450694

### CODE REVIEW

File: provisioning/tools-versions.yml [hunk 1]
```diff
@@ -1,19 +1,21 @@
 azurecli_version: "2.37.0"
+chocolatey_cigwin_version: "3.3.4.20220408"
+chocolatey_make_version: "4.3"
+chocolatey_windows_powershell_version: "5.1.14409.20180811"
```
Comment (lemeurherve) [inline provisioning/tools-versions.yml:4]:
```suggestion
```
Not using it

File: provisioning/tools-versions.yml [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
 jdk17_version: "17.0.3+7"
 jdk8_version: "8u332-b09"
 jq_version: "1.6"
 jxreleaseversion_version: "2.5.2"
 maven_version: "3.8.6"
-chocolatey_version: "1.1.0"
-chocolatey_make_version: "4.3"
-chocolatey_cigwin_version: "3.3.4.20220408"
 openssh_authorized_keys_url: "https://raw.githubusercontent.com/jenkins-infra/aws/main/ec2_agents_authorized_keys"
+powershell_core_version: "7.2.5"
```
Comment (lemeurherve) [inline provisioning/tools-versions.yml:21]:
```suggestion
```
same

**Review by dduportal — COMMENTED** [review]
Make sense 👍 

Of course it assumes that we are always installing the "latest" version of powershell (or pwsh): am I right?

File: provisioning/windows-provision.ps1 [hunk 3]
```diff
@@ -320,6 +320,22 @@ foreach($k in $downloads.Keys) {
     }
 }
 
+# Special case for Powershell, we need to make sure powershell.exe and pwsh.exe are both available
+# On Windows Server, Windows Powershell 5.1 is installed by default (powershell.exe)
+# On nanoserver, Powershell Core 7 is installed by default (pwsh.ex)
+# https://docs.microsoft.com/en-us/powershell/scripting/whats-new/migrating-from-windows-powershell-51-to-powershell-7?view=powershell-7.2#using-powershell-7-side-by-side-with-windows-powershell-51
+Write-Output "= Ensure both Windows Powershell and Powershell Core are available"
+if ((Get-Host | Select-Object Version).Version.Major == 5) {
```
Comment (lemeurherve) [inline provisioning/windows-provision.ps1:328]:
```suggestion
if ((Get-Host | Select-Object Version).Version.Major -eq 5) {

```

File: provisioning/windows-provision.ps1 [hunk 4]
```diff
@@ -320,6 +320,22 @@ foreach($k in $downloads.Keys) {
...
+if ((Get-Host | Select-Object Version).Version.Major -eq 5) {
+    Write-Output "== Windows Powershell already present, installing Powershell Core..."
+    Invoke-Command {& "choco.exe" install pwsh --yes --no-progress --limit-output --fail-on-error-output;}
+} else {
+    Write-Output "== Powershell Core already present, installing Windows Powershell..."
+    Invoke-Command {& "choco.exe" install powershell --yes --no-progress --limit-output --fail-on-error-output;}
+}
+Write-Output "== Windows Powershell & Powershell Core sanity checks:"
+Invoke-Command {& "powershell.exe" -command "(Get-Host).Version"}
+Invoke-Command {& "pwsh.exe" -command "(Get-Host).Version"}
```
Comment (lemeurherve) [inline provisioning/windows-provision.ps1:337]:
```suggestion
Write-Output "== Ensure both Windows Powershell and Powershell Core are available"
if ((Get-Host | Select-Object Version).Version.Major -eq 5) {
    Write-Output "= Windows Powershell already present, installing Powershell Core..."
    Invoke-Command {& "choco.exe" install pwsh --yes --no-progress --limit-output --fail-on-error-output;}
} else {
    Write-Output "= Powershell Core already present, installing Windows Powershell..."
    Invoke-Command {& "choco.exe" install powershell --yes --no-progress --limit-output --fail-on-error-output;}
}
Write-Output "= Windows Powershell & Powershell Core sanity checks:"
Invoke-Command {& "powershell.exe" -command "(Get-Host).Version"}
Invoke-Command {& "pwsh.exe" -command "(Get-Host).Version"}
```

File: provisioning/windows-provision.ps1 [hunk 5]
```diff
@@ -320,6 +320,24 @@ foreach($k in $downloads.Keys) {
...
+    Invoke-Command {& "choco.exe" install pwsh --yes --no-progress --limit-output --fail-on-error-output;}
+    AddToPathEnv "C:\Program Files\PowerShell\7\"
+} else {
+    Write-Output "= Powershell Core already present, installing Windows Powershell..."
+    Invoke-Command {& "choco.exe" install powershell --yes --no-progress --limit-output --fail-on-error-output;}
+    AddToPathEnv "C:\Windows\System32\WindowsPowerShell\v1.0\"
+}
+Write-Output "= Windows Powershell & Powershell Core sanity checks:"
+Invoke-Command {& "powershell.exe" -command "(Get-Host).Version"}
+Invoke-Command {& "pwsh.exe" -command "(Get-Host).Version"}
```
Comment (lemeurherve) [inline provisioning/windows-provision.ps1:339]:
```suggestion
Invoke-Command {& "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -command "(Get-Host).Version"}
Invoke-Command {& "C:\Program Files\PowerShell\7\pwsh.exe" -command "(Get-Host).Version"}
```

File: provisioning/tools-versions.yml (hunk 1 above)
Comment (dduportal) [reply provisioning/tools-versions.yml:4]: yep, remove it if not used :)

File: provisioning/windows-provision.ps1 [hunk 6]
```diff
@@ -320,6 +320,24 @@ foreach($k in $downloads.Keys) {
...
+    Invoke-Command {& "choco.exe" install pwsh --yes --no-progress --limit-output --fail-on-error-output;}
+    AddToPathEnv "C:\Program Files\PowerShell\7\"
+} else {
+    Write-Output "= Powershell Core already present, installing Windows Powershell..."
+    Invoke-Command {& "choco.exe" install powershell --yes --no-progress --limit-output --fail-on-error-output;}
+    AddToPathEnv "C:\Windows\System32\WindowsPowerShell\v1.0\"
+}
+Write-Output "= Windows Powershell & Powershell Core sanity checks:"
+Invoke-Command {& "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -command "(Get-Host).Version"}
+Invoke-Command {& "C:\Program Files\PowerShell\7\pwsh.exe" -command "(Get-Host).Version"}
```
Comment (dduportal) [inline provisioning/windows-provision.ps1:339]: ❤️

Comment (lemeurherve) [reply provisioning/windows-provision.ps1:339]: I would have liked to call these without a complete path but didn't managed to get it working.
Tested on a VM where it works without it. 🤷

Comment (lemeurherve) [reply provisioning/windows-provision.ps1:339]: Tried a `refreshenv` in the VM after the `choco install` as recommanded, no luck. Exiting powershell and executing these commands in cmd then worked.
I didn't wanted to loose too much time on this as they won't be called just after this provision script.

**Pull request merged**